pkg/kube/controllers/quarksstatefulset/active_passive_reconciler.go

```diff
@@ -0,0 +1,352 @@
+	return reconcile.Result{RequeueAfter: periodSeconds}, nil
+}
+
+func (r *ReconcileStatefulSetActivePassive) defineActiveContainer(ctx context.Context, container string, pods *corev1.PodList, qSts *qstsv1a1.QuarksStatefulSet) (err error) {
+	var p *corev1.Pod
+	labelledPods := r.getActiveLabels(pods, container)
+
+	probeCmd := qSts.Spec.ActivePassiveProbe[container].Exec.Command
+
+	// switch depending on pods with active label
+	switch pl := len(labelledPods); {
+	// Nothing have been labeled yet
+	case pl == 0:
+		ctxlog.WithEvent(qSts, "active-passive").Debugf(
+			ctx,
+			"there is no active pod",
+		)
+
+		p, err = r.getActivePod(pods, container, probeCmd)
+		if err != nil {
+			return err
+		}
+
+		podLabels := p.GetLabels()
+		if podLabels == nil {
+			podLabels = map[string]string{}
+
+		}
+
+		podLabels[qstsv1a1.LabelActiveContainer] = "active"
+
+		err = r.client.Update(r.ctx, p)
+		if err != nil {
+			return err
+		}
+		ctxlog.WithEvent(qSts, "active-passive").Debugf(
+			ctx,
+			"pod %s promoted to active",
+			p.Name,
+		)
+	case pl > 0:
+		for _, p := range labelledPods {
+			ctxlog.WithEvent(qSts, "active-passive").Debugf(
+				ctx,
+				"validating probe in active pod: %s",
+				p.Name,
+			)
+			if probeSucceeded, _ := r.execContainerCmd(&p, container, probeCmd); !probeSucceeded {
+				l := p.GetLabels()
+				delete(l, qstsv1a1.LabelActiveContainer)
+				err = r.client.Update(r.ctx, &p)
+				if err != nil {
+					return err
+				}
+				ctxlog.WithEvent(qSts, "active-passive").Debugf(
+					ctx,
+					"pod %s promoted to passive",
+					p.Name,
+				)
+			}
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileStatefulSetActivePassive) getActivePod(pods *corev1.PodList, container string, cmd []string) (*corev1.Pod, error) {
+	for _, pod := range pods.Items {
+		if succeed, _ := r.execContainerCmd(&pod, container, cmd); succeed && podutil.IsPodReady(&pod) {
+			return &pod, nil
+		}
+	}
+	return nil, errors.Errorf("executing cmd: %v, in container: %s, failed\n", cmd, container)
+}
+
+func (r *ReconcileStatefulSetActivePassive) getActiveLabels(pods *corev1.PodList, container string) []corev1.Pod {
+	var labelledPods []corev1.Pod
+	for _, pod := range pods.Items {
+		labels := pod.GetLabels()
+		if _, found := labels[qstsv1a1.LabelActiveContainer]; found {
+			labelledPods = append(labelledPods, pod)
+		}
+	}
+	return labelledPods
+}
+
+func (r *ReconcileStatefulSetActivePassive) execContainerCmd(pod *corev1.Pod, container string, command []string) (bool, error) {
+	req := r.kclient.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(pod.Name).
+		Namespace(pod.Namespace).
+		SubResource("exec").
+		VersionedParams(&corev1.PodExecOptions{
+			Container: container,
+			Command:   command,
+			Stdin:     true,
+			Stdout:    true,
+			Stderr:    false,
+		}, clientscheme.ParameterCodec)
+
+	executor, err := remotecommand.NewSPDYExecutor(r.kubeconfig, "POST", req.URL())
```

Comment:
is this the only way? the kubeconfig part looks strange

Reply:
maybe you got confuse, its actually a rest.Config